Unittest

ValeriePineauNoel opened the pull request:
Mostly tested the shift estimation. 
- 2 methods to tests, phase cross-correlation and FFT convolution. 
- Some tests DON'T pass and it's okay. FFT convolution doesn't seem to estimate the shift as well as the phase cross-correlation. I might work more on that eventually, but the tests are there to verify the performances of both methods at least.

ValeriePineauNoel committed:
Testing create black image

ValeriePineauNoel committed:
Add tests to the code.

Most tests are written. I might add some eventually.

FFTconvolution doesn't work very well to estimate the shift between two images. PCC have better results for now with intensity-corrected images.
Better to use PCC for now.

ValeriePineauNoel committed:
Little bit of corrections here and there

GabGabG reviewed (changes requested):
2/4 tests are failing on my computer:

- `test_PCCAndFFTConvolutionShifts`:
 - `test_estimateShift`
 
![image](https://user-images.githubusercontent.com/25515426/182953278-96c12e8d-b8ad-41bc-ba62-48fa3308a74e.png)
![image](https://user-images.githubusercontent.com/25515426/182953338-68667fe0-5a80-4aa7-8fe9-118c2b922acd.png)



As a general comment, I'm not sure I would use sub tests. To me it makes it messy to understand where exactly is the failure, like in the previous pictures: there is a lot of text. Also, there are some repetitions in your code. For example, `test_estimateShift` has pairs of subtests where only the assertion is different, perhaps you could use set ups (either before tests or before instantiating the class https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp)

Finally, in my opinion, the tests are well done (you test one concept at a time). I say "Request changes" because the tests are failing on my side. Other than that, it is up to you!

Edit: just saw your second "point" in the PR message, my bad! In that case where they are "ok" to fail, you could use skips so that they are not done https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures

GabGabG reviewed (approved):
Since some tests are expected to fail, I approve this PR, but please read my comments.

JLBegin commented on tests.py:
```diff
@@ -0,0 +1,130 @@
+pathTestImages = os.getcwd() + "/tests/testDataset"
+tileDim = [2,3] #[x,y]
+# test images are already intensity adjusted. 
+stitcherPCC = stitch.Stitching(sourceDir=pathTestImages, tileD=tileDim, imageSize=[1024,512], isIntensityCorrection=False, isMirrored=True, isFlipped=False)
+stitcherConv = stitch.Stitching(sourceDir=pathTestImages, tileD=tileDim, imageSize=[1024,512], isIntensityCorrection=False, shiftEstimation="FFTConvolution", isMirrored=True, isFlipped=False)
```
Attention tu réutilise la meme instance dans plusieurs tests. Un test doit être indépendant (start from scratch à chaque fois). Ca ne crée pas trop de problème en ce moment car ton code n'est pas orienté objet, mais si une méthode commence à jouer sur l'état de l'objet un jour ca risque de faire échouer d'autres tests de manière imprédictible et difficile à corriger. Je metterais cest objets dans un TestClass.setUp() (exécuté entre chaque test function).

JLBegin commented on tests.py:
```diff
@@ -0,0 +1,130 @@
+stitcherPCC = stitch.Stitching(sourceDir=pathTestImages, tileD=tileDim, imageSize=[1024,512], isIntensityCorrection=False, isMirrored=True, isFlipped=False)
+stitcherConv = stitch.Stitching(sourceDir=pathTestImages, tileD=tileDim, imageSize=[1024,512], isIntensityCorrection=False, shiftEstimation="FFTConvolution", isMirrored=True, isFlipped=False)
+
+class TestStitching(unittest.TestCase):
+	def test_calculateShiftPCC_horizontalShift(self):
+		with self.subTest("Horizontal shift estimation with PCC returns a list."):
+			shift = stitcherPCC.calculate_shift_PCC(imageRef1=0, imageRef2=1)
+			self.assertIsInstance(shift, list)
```
Les subtests ne donne pas le meilleur traceback, cest préférable de mettre tout le contexte dans le nom de la fonction de test. Tu peux compartimenter en utilisant plusieurs classes de tests. Par exemple, de manière grossière dans ton cas j'irais peut-être transformer plusieurs fonctions en une classe et chaque subTest en une fonction. 

_Note plus avancé_: ca cest aussi un indicateur que ta classe Stitching a trop de responsabilité et qu'elle devrait composer d'autres petites classes. 

_Note plus avancé de réflexion qui ne sert pas à grand chose dans ton cas, mais bon ca me tente de parler_: le but d'un test idéal est de documenter le outcome et non le output. Quand il est écrit "returns" dans le nom du test cest un code smell qui indique que tu ne test pas le comportement attendu. Pas besoin de tester le format du output ("cest une liste de dimension X", etc. ). Mais bon, cette règle fonctionne surtout dans un cadre orienté objet ou avec des fonctions d'action qui applique un changement sur l'entrée. Dans ton cas tes fonctions sont plus des propriétés comme getShift(image) ou getShape(image) sans comportement, et non des méthodes d'action comme rotate(image) ou shift(image), alors je ne sais pas trop quoi dire. Règle générale c'est rare quon a des propriétés publiques et même là c'est rare quon doit les tester explicitement. Pour une classe qui s'appelle Stitching je m'attenderais à avoir environ une seule méthode publique nommé 'stitch(images)'.

JLBegin reviewed (approved):
Les tests semblent bien fonctionner, autre que les tests qui fail déjà pointé par Gab (tu pourrais aussi changer le titre du test pour documenter que tel méthode ne permet pas dobtenir la bonne valeur, et, de même, changer l'assertion pour un assertNotEqual), et le contenu semble assez bien tester ce que le contexte dit qu'il test. Je vais juste indiquer des pistes d'amélioration.

ValeriePineauNoel committed:
Add and test sobel filter

ValeriePineauNoel committed:
Tried to estimate the vertical and horizontal shifts with beads and apply on the biological samples.

Didn't work. Even if the images were acquired on the same day with the same "overlap" and the same zoom factor, for some reasons, the shifts applied on the beads is not the same than the shift from the biological sample.

ValeriePineauNoel committed:
added a function that uses the position of the stage when the image was acquired to calculate the shift between two images

In Nirvana, the position of the stage is an information found in the file name, so it would be possible to calculate the shift by using this. However, this does not consider the small lateral shift caused by a misalignment of the stage.

Valeriouus commented:
NOTE à MOI-MÊME : 
J'ai oublié de me créer une nouvelle branche, donc j'ai fait quelques tests ce matin sur le code que j'ai ajouté ici (oups). 
Voici ce que j'ai essayé pour que mes images d'échantillons biologiques stitch comme du monde (parce que les billes marchent! mais pas les échantillons bio):
- Estimer le shift (overlap + déplacement latéral) avec l'échantillon de bille qui a été acquis le même jour avec le même overlap, puis moyenner pour obtenir un seul shift plus général. Ce shift peut être utiliser pour placer les images d'échantillons biologiques qui ont été prises le même jour avec le même overlap. Vu que je change l'angle du goniomètre chaque fois que j'image quelque chose de différent, je pense que le petit déplacement latéral est modifié. Ce n'est donc pas possible d'estimer le déplacement latéral une fois et de l'appliquer pour toutes mes images; il faudrait l'estimer à chaque tile. 
- Calculer le shift selon la position du stage dans le nom du fichier. Nirvana met dans le nom du fichier la position du stage a laquelle l'image a été acquise. En ayant extraire cette information pour deux images, on peut calculer le shift. Par contre, ce shift ne considère que l'overlap, et non le déplacement latéral. Il faudrait quand même trouver une façon d'estimer le déplacement latéral de façon computationnelle mais, au moins, on peut les placer selon le overlap.
